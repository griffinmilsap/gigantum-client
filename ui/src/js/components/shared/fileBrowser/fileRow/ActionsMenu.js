// vendor
import React, { Component, Fragment } from 'react';
import classNames from 'classnames';
// assets
import './ActionsMenu.scss';

export default class ActionsMenu extends Component {
  state = {
    popupVisible: false,
  }

  /**
  *  LIFECYCLE MEHTODS START
  */
  componentWillMount() {
    window.removeEventListener('click', this._closePopup);
  }

  componentDidMount() {
    window.addEventListener('click', this._closePopup);
  }
  /**
  *  LIFECYCLE MEHTODS END
  */

  /**
  *  @param {Object} event
  *  closes popup when clicking
  *  @return {}
  */
  _closePopup = (evt) => {
    const { props, state } = this;
    const node = this[props.edge.node.id];
    if (state.popupVisible && node && !node.contains(evt.target)) {
      this.setState({ popupVisible: false });
    }
  }

  /**
  *  @param {Obect} evt
  *  @param {boolean} popupVisible - boolean value for hiding and showing popup state
  *  triggers favoirte unfavorite mutation
  *  @return {}
  */
  _togglePopup = (evt, popupVisible) => {
    if (!popupVisible) {
      evt.stopPropagation(); // only stop propagation when closing popup
    }
    this.setState({ popupVisible });
  }

  /**
  *  @param {Object} data - event from clicking delete button
  *  triggers DeleteLabbookFileMutation
  *  @return {}
  */
  _getEdges = (data) => {
    const edges = [data.edge];
    const getEdges = (childData) => {
      Object.keys(childData).forEach((name) => {
        edges.push(childData[name].edge);
        if (childData[name].children) {
          getEdges(childData[name].children);
        }
      });
      return edges;
    };

    if (data.children) {
      getEdges(data.children);
    }
    return edges;
  }

  /**
  *  @param {event} evt - event from clicking delete button
  *  triggers DeleteLabbookFileMutation
  *  @return {}
  */
  _triggerDeleteMutation = (evt) => {
    const { props } = this;
    const edges = props.fileData
      ? this._getEdges(props.fileData)
      : [props.edge];
    const deleteFileData = {
      filePaths: [props.edge.node.key],
      edges,
    };

    props.mutations.deleteLabbookFiles(deleteFileData, () => {});

    this._togglePopup(evt, false);
  }

  /**
  *  @param {Object} node - Dom object to be assigned as a ref
  *  set wrapper ref
  *  @return {}
  */
  _setWrapperRef = (node) => {
    const { props } = this;
    this[props.edge.node.id] = node;
  }

  /**
  *  @param {} -
  *  shows folder if it exists
  *  @return {}
  */
  _addFolderVisibile = () => {
    const { props } = this;
    if (props.folder) {
      props.addFolderVisible(true);
    }
  }

  render() {
    const { props, state } = this;
    const disableButtons = props.section === 'data' && (!props.edge.node.isLocal || (props.folder && !props.isLocal));
    const deleteTooltip = disableButtons ? 'Must download before deleting' : 'Delete';
    const renameTooltip = disableButtons ? 'Must download before renaming' : 'Rename';
    const isUntrackedDirectory = (props.edge.node.key === 'untracked/')
      && props.folder
      && (props.section !== 'data');
    // declare css here
    const popupCSS = classNames({
      ActionsMenu__popup: true,
      hidden: !state.popupVisible,
      Tooltip__message: true,
    });
    const deleteCSS = classNames({
      'ActionsMenu__item Btn Btn--fileBrowser Btn__delete-secondary Btn--round': true,
      'Tooltip-data Tooltip-data--small': !state.popupVisible,
      'ActionsMenu__popup-visible': state.popupVisible,
    });
    const folderCSS = classNames({
      'ActionsMenu__item Btn Btn--fileBrowser Tooltip-data Tooltip-data--small Btn--round': true,
      Btn__addFolder: true,
      'visibility-hidden': !props.folder,
    });

    return (
      <div
        className="ActionsMenu"
        key={`${props.edge.node.id}-action-menu}`}
        ref={this._setWrapperRef}
      >
        <button
          onClick={() => { this._addFolderVisibile(); }}
          className={folderCSS}
          data-click-id="addFolder"
          data-tooltip="Add Subfolder"
          type="button"
        />
        { !isUntrackedDirectory
            && (
              <Fragment>
                <div className="relative">
                  <button
                    className={deleteCSS}
                    data-tooltip={deleteTooltip}
                    onClick={(evt) => { this._togglePopup(evt, true); }}
                    disabled={disableButtons}
                    type="button"
                  />
                  <div className={popupCSS}>
                    <div className="Tooltip__pointer" />
                    <p>Are you sure?</p>
                    <div className="flex justify--space-around">
                      <button
                        className="File__btn--round File__btn--cancel"
                        onClick={(evt) => { this._togglePopup(evt, false); }}
                        type="button"
                      />
                      <button
                        className="File__btn--round File__btn--add"
                        onClick={(evt) => { this._triggerDeleteMutation(evt); }}
                        type="button"
                      />
                    </div>
                  </div>
                </div>
                <button
                  disabled={disableButtons}
                  onClick={() => { props.renameEditMode(true); }}
                  className="ActionsMenu__item Btn Btn--fileBrowser Btn__rename Tooltip-data Tooltip-data--small Btn--round"
                  data-tooltip={renameTooltip}
                  type="button"
                />
              </Fragment>
            )
          }
      </div>
    );
  }
}
