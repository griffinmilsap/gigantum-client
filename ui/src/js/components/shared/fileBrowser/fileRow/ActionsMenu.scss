@import '~Styles/_imports.scss';
.FileBrowser{
  .ActionsMenu{
      right: 20px;

      padding: 2px 20px 2px 2px;

      width: 180px;

      @include flex(flex-end, row);
      .Tooltip-data:before {
        left: calc(50% - 7px);
      }

  }
  .ActionsMenu__item {
      position: relative;

      width: 20px;
      height: 20px;

      margin: 0 5px;

      cursor: pointer;


      &--delete {
          background: url('~Images/icons/trash.svg') no-repeat 50% 50%;
          background-size: 20px;

          position: relative;

          &:hover{
              //change color here
              background: url('~Images/icons/trash-azure.svg') no-repeat 50% 50%;
              background-size: 20px;
          }
      }

      &--rename {
          background: url('~Images/icons/rename.svg') no-repeat 50% 50%;
          background-size: 20px;

          &:hover{
              //change color here
              background: url('~Images/icons/rename-azure.svg') no-repeat 50% 50%;
              background-size: 20px;
          }
      }

      &--duplicate {
          background: url('~Images/icons/copy-light.svg') no-repeat 50% 50%;
          background-size: 24px;

          &:hover{
              //change color here
              background: url('~Images/icons/copy-light.svg') no-repeat 50% 50%;
              background-size: 24px;
          }
      }

      &--favorite-on {
          background: url('~Images/icons/star__on.svg') no-repeat 50% 50%;
          background-size: 20px;

          &:hover {
              //change color here
              background: url('~Images/icons/star__on.svg') no-repeat 50% 50%;
              background-size: 20px;
          }
      }

      &--favorite-off {
          background: url('~Images/icons/star__off.svg') no-repeat 50% 50%;
          background-size: 20px;

          &:hover {
              //change color here
              background: url('~Images/icons/star__off.svg') no-repeat 50% 50%;
              background-size: 20px;
          }
      }

      &--AddSubfolder {
          background: url('~Images/icons/folder-add.svg') no-repeat 50% 50%;
          background-size: 20px;
          margin-right: 5px;

          &:hover {
              //change color here
              background: url('~Images/icons/folder-add.svg') no-repeat 50% 50%;
              background-size: 20px;
          }
      }

      &--menu {
          width: 50px;

          cursor: pointer;

          background: transparent url('~Images/icons/ellipsis.svg') no-repeat 50% 50%;
          background-size: 60px 30px;

          &:hover {
              //change color here
              background: transparent url('~Images/icons/ellipsis.svg') no-repeat 50% 50%;
              background-size: 65px 35px;
          }
      }

  }

.ActionsMenu__popup-visible:hover:after{
  display: none;
  color: transparent;
  background: transparent;
}

.ActionsMenu__popup {
  width: 140px;
  position: absolute;
  top: 34px;
  right: -60px;
  text-align: center;
  z-index: index($elements, popup);
  p{
      margin-top: 0;
  }

  .Tooltip__pointer {
    top: -10px;
    right: 55px;
  }
}
}
