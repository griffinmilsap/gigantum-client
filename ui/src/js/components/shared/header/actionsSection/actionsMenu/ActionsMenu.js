// @flow
// vendor
import React, { Component } from 'react';
import classNames from 'classnames';
import uuidv4 from 'uuid/v4';
import { connect } from 'react-redux';
// utilities
import JobStatus from 'JS/utils/JobStatus';
// mutations
import ExportLabbookMutation from 'Mutations/ExportLabbookMutation';
import ExportDatasetMutation from 'Mutations/ExportDatasetMutation';
import SyncLabbookMutation from 'Mutations/branches/SyncLabbookMutation';
import SyncDatasetMutation from 'Mutations/branches/SyncDatasetMutation';
import BuildImageMutation from 'Mutations/container/BuildImageMutation';
// store
import {
  setErrorMessage,
  setWarningMessage,
  setInfoMessage,
  setMultiInfoMessage,
} from 'JS/redux/actions/footer';
import store from 'JS/redux/store';
import { updateTransitionState } from 'JS/redux/actions/labbook/labbook';
import { setContainerMenuWarningMessage, setContainerMenuVisibility } from 'JS/redux/actions/labbook/environment/environment';
// queries
import UserIdentity from 'JS/Auth/UserIdentity';
import LinkedLocalDatasetsQuery from 'Components/shared/header/actionsSection/queries/LinkedLocalDatasetsQuery';
// components
import CreateBranch from 'Components/shared/modals/CreateBranch';
import Tooltip from 'Components/common/Tooltip';
import LoginPrompt from 'Components/shared/modals/LoginPrompt';
import VisibilityModal from 'Components/shared/modals/VisibilityModal';
import DeleteLabbook from 'Components/shared/modals/DeleteLabbook';
import DeleteDataset from 'Components/shared/modals/DeleteDataset';
// assets
import './ActionsMenu.scss';

type Props = {
  auth: {
    renewToken: Function,
  },
  defaultRemote: string,
  description: string,
  history: Object,
  isExporting: boolean,
  isLocked: bool,
  name: string,
  owner: string,
  remoteUrl: string,
  sectionType: string,
  setExportingState: Function,
  setSyncingState: Function,
  toggleBranchesView: Function,
  visibility: string,
};


class ActionsMenu extends Component<Props> {
  state = {
    addNoteEnabled: false,
    isValid: true,
    createBranchVisible: false,
    showLoginPrompt: false,
    exporting: false,
    deleteModalVisible: false,
    publishDisabled: false,
    justOpened: true,
    setPublic: false,
    syncWarningVisible: false,
    publishWarningVisible: false,
    visibilityModalVisible: false,
    remoteUrl: this.props.remoteUrl,
    owner: this.props.owner,
    name: this.props.owner,
  };


  /**
   * attach window listener evetns here
  */
  componentDidMount() {
    window.addEventListener('click', this._closeMenu);
  }

  /**
   * detach window listener evetns here
  */
  componentWillUnmount() {
    window.removeEventListener('click', this._closeMenu);
  }

  /**
    @param {event} evt
    closes menu
  */
  _closeMenu = (evt) => {
    const {
      menuOpen,
      syncWarningVisible,
      publishWarningVisible,
    } = this.state;
    const isActionsMenu = (evt.target.className.indexOf('ActionsMenu') > -1)
      || (evt.target.className.indexOf('CollaboratorsModal') > -1)
      || (evt.target.className.indexOf('ActionsMenu__message') > -1)
      || (evt.target.className.indexOf('TrackingToggle') > -1);

    if (!isActionsMenu && menuOpen) {
      this.setState({ menuOpen: false, justOpened: true });
    }

    if (
      (evt.target.className.indexOf('ActionsMenu__btn--sync') === -1)
      && syncWarningVisible
    ) {
      this.setState({ syncWarningVisible: false });
    }

    if (
      (evt.target.className.indexOf('ActionsMenu__btn--remote') === -1)
      && publishWarningVisible
    ) {
      this.setState({ publishWarningVisible: false });
    }
  }

  /**
    @param {string} value
    sets state on createBranchVisible and toggles modal cover
  */
  _toggleModal = (value) => {
    this.setState((state) => {
      const inverseValue = !state[value];
      return {
        [value]: inverseValue,
      };
    });
  }

  /**
  *  @param {}
  *  toggles open menu state
  *  @return {string}
  */
  _toggleMenu = () => {
    this.setState((state) => {
      const menuOpen = !state.menuOpen;
      return ({ menuOpen });
    });

    if (!this.state.menuOpen) {
      setTimeout(() => {
        this.setState({ justOpened: false });
      }, 500);
    } else {
      this.setState({ justOpened: true });
    }
  }

  /**
  *  @param {}
  *  remounts collaborators by updating key
  *  @return {}
  */
  _remountCollab = () => {
    this.setState({ collabKey: uuidv4() });
  }

  /**
  *  @param {string, boolean} action, containerRunning
  *  displays container menu message
  *  @return {}
  */
  _showContainerMenuMessage = (action, containerRunning) => {
    const dispatchMessage = containerRunning ? `Stop Project before ${action}. \n Be sure to save your changes.` : `Project is ${action}. \n Please do not refresh the page.`;

    this.setState({ menuOpen: false });

    setContainerMenuWarningMessage(dispatchMessage);
  }

  /**
  *  @param {Boolean} pullOnly
  *  pushes code to remote
  *  @return {string}
  */
  _sync = (pullOnly) => {
    const {
      auth,
      owner,
      name,
      isExporting,
      setSyncingState,
      sectionType,
    } = this.props;

    if (isExporting) {
      this.setState({ syncWarningVisible: true });
    } else {
      const { status } = store.getState().containerStatus;
      this.setState({ pullOnly });

      if (owner !== 'gigantum-examples') {
        this.setState({ menuOpen: false });
      }

      if (
        (status === 'Stopped')
        || (status === 'Rebuild')
        || (sectionType !== 'labbook')
      ) {
        const id = uuidv4();
        const self = this;

        this._checkSessionIsValid().then((response) => {
          if (navigator.onLine) {
            if (response.data && response.data.userIdentity) {
              if (response.data.userIdentity.isSessionValid) {
                const failureCall = (errorMessage) => {
                  setSyncingState(false);
                  if (errorMessage.indexOf('Merge conflict') > -1) {
                    self._toggleSyncModal();
                  }
                };

                const successCall = () => {
                  setSyncingState(false);
                  if (sectionType === 'labbook') {
                    BuildImageMutation(
                      owner,
                      name,
                      false,
                      (response, error) => {
                        if (error) {
                          console.error(error);
                          const messageData = {
                            id,
                            message: `ERROR: Failed to build ${name}`,
                            isLast: null,
                            error: true,
                            messageBody: error,
                          };
                          setMultiInfoMessage(owner, name, messageData);
                        }
                      },
                    );
                  }

                  setContainerMenuVisibility(false);
                };
                if (sectionType === 'labbook') {
                  LinkedLocalDatasetsQuery.getLocalDatasets({
                    owner,
                    name,
                  }).then((res) => {
                    const localDatasets = res.data
                      && res.data.labbook.linkedDatasets.filter(linkedDataset => linkedDataset.defaultRemote && linkedDataset.defaultRemote.slice(0, 4) !== 'http');

                    if (localDatasets.length === 0) {
                      setSyncingState(true);

                      this._showContainerMenuMessage('syncing');
                      SyncLabbookMutation(
                        owner,
                        name,
                        null,
                        pullOnly,
                        successCall,
                        failureCall,
                        (error) => {
                          if (error) {
                            failureCall(error);
                          }
                        },
                      );
                    } else {
                      this.setState((state) => {
                        const publishDatasetsModalVisible = !state.publishDatasetsModalVisible;
                        return {
                          localDatasets,
                          publishDatasetsModalVisible,
                          publishDatasetsModalAction: 'Sync',
                        };
                      });
                    }
                  });
                } else {
                  setSyncingState(true);

                  this._showContainerMenuMessage('syncing');
                  SyncDatasetMutation(
                    owner,
                    name,
                    false,
                    successCall,
                    failureCall,
                    (error) => {
                      if (error) {
                        failureCall(error);
                      }
                    },
                  );
                }
              } else {
                auth.renewToken(true, () => {
                  self.setState({ showLoginPrompt: true });
                }, () => {
                  self._sync(pullOnly);
                });
              }
            }
          } else {
            self.setState({ showLoginPrompt: true });
          }
        });
      } else {
        this.setState({ menuOpen: false });

        setContainerMenuWarningMessage('Stop Project before syncing. \n Be sure to save your changes.');
      }
    }
  }

  /**
  *  @param {}
  *  shows collaborators warning if user is not owner
  *  @return {}
  */
  _showCollaboratorsWarning = () => {
    const { owner, name } = this.props;
    const username = localStorage.getItem('username');

    if (owner !== username) {
      setWarningMessage(owner, name, `Only ${owner} can add and remove collaborators in this labbook.`);
    }
  }

  /**
  *  @param {}
  *  returns UserIdentityQeury promise
  *  @return {promise}
  */
  _checkSessionIsValid = () => {
    return (UserIdentity.getUserIdentity());
  }

  /**
  *  @param {}
  *  closes login prompt modal
  *  @return {}
  */
  _closeLoginPromptModal = () => {
    this.setState({ showLoginPrompt: false });
  }

  /**
  *  @param {}
  *  copies remote
  *  @return {}
  */
  _copyRemote = () => {
    const {
      owner,
      name,
    } = this.state;
    const copyText = document.getElementById('ActionsMenu-copy');
    copyText.select();

    document.execCommand('Copy');

    setInfoMessage(owner, name, `${copyText.value} copied!`);
  }

  /**
  *  @param {jobKey}
  *  polls jobStatus for export job message
  *  updates footer with a message
  *  @return {}
  */
  _jobStatus = (jobKey) => {
    const {
      setExportingState,
      owner,
      name,
    } = this.props;

    JobStatus.getJobStatus(owner, name, jobKey).then((data) => {
      setExportingState(false);
      updateTransitionState(owner, name, '');

      if (data.jobStatus.result) {
        setInfoMessage(owner, name, `Export file ${data.jobStatus.result} is available in the export directory of your Gigantum working directory.`);
      }

      this.setState({ exporting: false });
    }).catch((error) => {
      updateTransitionState(owner, name, '');
      console.log(error);

      setExportingState(false);

      const errorArray = [{ message: 'Export failed.' }];

      setErrorMessage(owner, name, `${name} failed to export `, errorArray);

      this.setState({ exporting: false });
    });
  }

  /**
  *  @param {}
  *  runs export mutation if export has not been downloaded
  *  @return {}
  */
  _exportLabbook = () => {
    const {
      sectionType,
      setExportingState,
      owner,
      name,
      isLocked,
    } = this.props;

    if (!isLocked) {
      this.setState({
        exporting: true,
        menuOpen: false,
      });
      const exportType = (sectionType === 'dataset') ? 'Dataset' : 'Project';

      setInfoMessage(owner, name, `Exporting ${name} ${exportType}`);
      updateTransitionState(owner, name, 'Exporting');

      setExportingState(true);
      if (sectionType !== 'dataset') {
        ExportLabbookMutation(
          owner,
          name,
          (response, error) => {
            if (response.exportLabbook) {
              this._jobStatus(response.exportLabbook.jobKey);
            } else {
              console.log(error);

              setExportingState(false);

              setErrorMessage(owner, name, 'Export Failed', error);
            }
          },
        );
      } else {
        ExportDatasetMutation(
          owner,
          name,
          (response, error) => {
            if (response.exportDataset) {
              this._jobStatus(response.exportDataset.jobKey);
            } else {
              console.log(error);

              setExportingState(false);

              setErrorMessage(owner, name, 'Export Failed', error);
            }
          },
        );
      }
    } else {
      this._showContainerMenuMessage('exporting', true);
    }
  }

  /**
  *  @param {}
  *  toggle stat and modal visibility
  *  @return {}
  */
  _toggleDeleteModal = () => {
    this.setState((state) => {
      const deleteModalVisible = !state.deleteModalVisible;
      return {
        deleteModalVisible,
      };
    });
  }

  /**
  *  @param {}
  *  sets menu
  *  @return {}
  */
  _mergeFilter = () => {
    const { toggleBranchesView } = this.props;
    if (store.getState().containerStatus.status !== 'Running') {
      toggleBranchesView(true, true);

      this.setState({ menuOpen: false });

      window.scrollTo(0, 0);
    } else {
      this._showContainerMenuMessage('merging branches', true);
    }
  }

  /**
  *  @param {}
  *  sets menu
  *  @return {}
  */
  _switchBranch = () => {
    const { toggleBranchesView } = this.props;
    const { status } = store.getState().containerStatus;

    if (status !== 'Running') {
      window.scrollTo(0, 0);

      toggleBranchesView(true, false);

      this.setState({ menuOpen: false });
    } else {
      this._showContainerMenuMessage('switching branches', true);
    }
  }

  /**
  *  @param {string} modal
  *  passes modal to toggleModal if container is not running
  *  @return {}
  */
  _handleToggleModal = (modal) => {
    let action = '';

    if (store.getState().containerStatus.status !== 'Running') {
      this._toggleModal(modal);
    } else {
      switch (modal) {
        case 'createBranchVisible':
          action = 'creating branches';
          break;
        default:
          break;
      }

      this._showContainerMenuMessage(action, true);
    }
  }

  /**
  *  @param {}
  *  resets state after publish
  *  @return {}
  */
  _resetState = () => {
    this.setState({
      remoteUrl: '',
      showLoginPrompt: true,
    });
  }

  /**
  *  @param {}
  *  resets state after publish
  *  @return {}
  */
  _resetPublishState = (publishDisabled) => {
    this.setState({
      menuOpen: false,
      publishDisabled,
    });
  }

  /**
  *  @param {}
  *  resets state after publish
  *  @return {}
  */
  _setRemoteSession = () => {
    const { owner, name } = this.props;
    this.setState({
      addedRemoteThisSession: true,
      remoteUrl: `https://gigantum.com/${owner}/${name}`,
    });
  }


  render() {
    const {
      name,
      owner,
      sectionType,
      defaultRemote,
      history,
      auth,
      visibility,
      description,
      isLocked,
    } = this.props;
    const {
      menuOpen,
      showLoginPrompt,
      deleteModalVisible,
      visibilityModalVisible,
      remoteUrl,
      justOpened,
      exporting,
    } = this.state;
    const deleteText = (sectionType === 'labbook') ? 'Delete Project' : 'Delete Dataset';
    // declare css here
    const branchMenuCSS = classNames({
      'ActionsMenu__menu--animation': justOpened, // this is needed to stop animation from breaking position flow when collaborators modal is open
      hidden: !menuOpen,
      'ActionsMenu__menu box-shadow': true,
    });

    return (
      <div className="ActionsMenu flex flex--column'">

        { showLoginPrompt
          && (
            <LoginPrompt closeModal={this._closeLoginPromptModal} />
          )
        }

        { (deleteModalVisible && (sectionType === 'labbook'))
          && (
            <DeleteLabbook
              handleClose={() => this._toggleDeleteModal()}
              remoteAdded={defaultRemote}
              history={history}
              name={name}
              owner={owner}
              remoteDelete={false}
            />
          )
        }

        { (deleteModalVisible && (sectionType === 'dataset'))
          && (
            <DeleteDataset
              handleClose={() => this._toggleDeleteModal()}
              remoteAdded={defaultRemote}
              history={history}
              name={name}
              owner={owner}
            />
          )
        }

        { visibilityModalVisible
          && (
          <VisibilityModal
            sectionType={sectionType}
            owner={owner}
            name={name}
            auth={auth}
            toggleModal={this._toggleModal}
            buttonText="Save"
            header="Change Visibility"
            modalStateValue="visibilityModalVisible"
            checkSessionIsValid={this._checkSessionIsValid}
            resetState={this._resetState}
            visibility={visibility}
          />
          )
        }

        <CreateBranch
          description={description}
          modalVisible={this.state.createBranchVisible}
          toggleModal={this._toggleModal}
        />

        <button
          onClick={() => { this._toggleMenu(); }}
          className="ActionsMenu__btn Btn--last"
          type="button"
        />

        <div className={branchMenuCSS}>

          <ul className="ActionsMenu__list">

            <li className="ActionsMenu__item ActionsMenu__item--export">
              <button
                onClick={evt => this._exportLabbook(evt)}
                disabled={exporting || isLocked}
                className="ActionsMenu__btn--flat"
                type="button"
                data-tooltip="Cannot export Project while in use"
              >
                Export to Zip
              </button>

              <div
                className="Tooltip-data Tooltip-data--top-offset Tooltip-data--info"
                data-tooltip="Exports project zip file to your gignatum directory"
              />

            </li>


            <li className="ActionsMenu__item ActionsMenu__item--delete">

              <button
                onClick={() => this._toggleDeleteModal()}
                className="ActionsMenu__btn--flat"
                type="button"
              >
                {deleteText}
              </button>

            </li>

            { defaultRemote
              && (
              <li className={`ActionsMenu__item ActionsMenu__item--visibility-${visibility}`}>

                <button
                  onClick={() => this._toggleModal('visibilityModalVisible')}
                  className="ActionsMenu__btn--flat"
                  type="button"
                >
                  Change Visibility
                </button>

              </li>
              )
            }
            { (remoteUrl || defaultRemote)
              && (
              <li className="ActionsMenu__item ActionsMenu__item--copy">
                <div className="ActionsMenu__item--label">Get Share URL</div>
                <div className="ActionsMenu__copyRemote">

                  <input
                    id="ActionsMenu-copy"
                    className="ActionsMenu__input"
                    defaultValue={`gigantum.com/${owner}/${name}`}
                    type="text"
                  />

                  <button
                    onClick={() => this._copyRemote()}
                    className="ActionsMenu__btn--copy fa fa-clone"
                    type="button"
                  />
                </div>
              </li>
              )
            }

          </ul>

        </div>

        <Tooltip section="actionMenu" />

      </div>
    );
  }
}

const mapStateToProps = state => state.packageDependencies;

const mapDispatchToProps = () => ({
  setContainerMenuWarningMessage,
});

export default connect(mapStateToProps, mapDispatchToProps)(ActionsMenu);
