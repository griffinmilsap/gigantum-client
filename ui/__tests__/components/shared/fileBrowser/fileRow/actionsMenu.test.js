// vendor
import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import relayTestingUtils from '@gigantum/relay-testing-utils';
// data
import codeData from 'Tests/components/labbook/code/__relaydata__/CodeBrowser.json';
// components
import ActionsMenu from 'Components/shared/fileBrowser/fileRow/ActionsMenu';

let edge = codeData.data.labbook.code.allFiles.edges[1]

const mutationData = {
  owner: 'uitest',
  labbookName: 'ui-test-project',
  section: 'code',
  connection: 'CodeBrowser_allFiles',
  favoriteConnection: 'CodeFavorite_favorites',
  parentId: codeData.data.labbook.id,
};

let file = {
  edge,
};

const fixtures = {
  closeLinkModal: jest.fn(),
  togglePopup: jest.fn(),
  addFolderVisible: jest.fn(),
  renameEditMode: jest.fn(),
  edge,
  mutationData,
  mutations: {
    moveLabbookFile: jest.fn(),
  },
  fileData: file,
  folder: false,
};


describe('ActionsMenu Component', () => {
  it('Renders a snapshot', () => {
    const wrapper = renderer.create(
      <ActionsMenu
        {...fixtures}
      />,
    );

    const tree = wrapper.toJSON();

    expect(tree).toMatchSnapshot();
  });
});
