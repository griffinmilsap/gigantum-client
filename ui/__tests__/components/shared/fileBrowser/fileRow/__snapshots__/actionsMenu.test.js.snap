// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActionsMenu Component Renders a snapshot 1`] = `
<div
  className="ActionsMenu"
  data-selenium-id="ActionsMenu"
>
  <div
    className="ActionsMenu__item Tooltip-data Tooltip-data--small ActionsMenu__item--AddSubfolder visibility-hidden"
    data-tooltip="Add Subfolder"
    onClick={[Function]}
  />
  <button
    className="ActionsMenu__item Btn__delete Btn--noShadow Tooltip-data Tooltip-data--small"
    data-tooltip="Delete"
    disabled={false}
    onClick={[Function]}
  >
    <div
      className="ActionsMenu__popup hidden Tooltip__message"
    >
      <div
        className="Tooltip__pointer"
      />
      <p>
        Are you sure?
      </p>
      <div
        className="flex justify--space-around"
      >
        <button
          className="File__btn--round File__btn--cancel"
          onClick={[Function]}
        />
        <button
          className="File__btn--round File__btn--add"
          onClick={[Function]}
        />
      </div>
    </div>
  </button>
  <button
    className="ActionsMenu__item Btn__rename Btn--noShadow Tooltip-data Tooltip-data--small"
    data-tooltip="Rename"
    disabled={false}
    onClick={[Function]}
  />
  <button
    className="ActionsMenu__item Tooltip-data Tooltip-data--small Btn__Favorite-off Btn--noShadow"
    data-tooltip="Favorite"
    onClick={[Function]}
  />
</div>
`;
